### ISSUE TITLE
when using layers.embedding, is_sparse=True is much slower than is_sparse=False

### ISSUE THREAD

**Issue by LihangLiu** [PaddlePaddle/Paddle#14861]
使用layers.embedding的时候，发现is_sparse=True 比 is_sparse=False 还慢很多。
代码如下，环境为fluid1.1。net函数中用到embedding。

    import paddle
    import paddle.fluid as fluid
    from paddle.fluid import layers
    from paddle.fluid.param_attr import ParamAttr
    import numpy as np
    import time

    def fluid_create_lod_tensor(array, lod, place):
        tensor = fluid.LoDTensor()
        tensor.set(np.array(array), place)
        tensor.set_lod(lod)
        return tensor

    def net():
        x = layers.data(name='x', shape=[-1, 1], dtype='int64')
        xavier_initializer = fluid.initializer.Xavier()
        param_clip = fluid.clip.GradientClipByValue(1.0)
        embed = layers.embedding(x,
                        size=[1000000, 16],
                        param_attr=ParamAttr(name='embed',
                                            initializer=xavier_initializer,
                                            gradient_clip=param_clip
                                            ),
                        is_sparse=False)
                        # is_sparse=True)
        out = layers.fc(embed, size=1)
        loss = layers.reduce_mean(layers.square(out))
        opt = fluid.optimizer.AdamOptimizer(
            learning_rate=0.01, beta1=0.9, beta2=0.999)
        opt.minimize(loss)
        return loss

                                    
    def train():
        myprogram = fluid.Program()
        with fluid.program_guard(myprogram):
            out = net()
        test_program = myprogram.clone(for_test=True)
        
        place = fluid.CPUPlace()
        exe = fluid.Executor(place)
        exe.run(fluid.default_startup_program())

        # data
        x_data = np.ones([10, 1]).astype('int64')
        x_lod = []
        x_tensor = fluid_create_lod_tensor(x_data, x_lod, place)

        time_list = []
        for _ in range(10):
            s = time.time()
            results = exe.run(myprogram, feed={'x': x_tensor}, fetch_list=[out])
            time_list.append(time.time() - s)
        print ('time', np.mean(time_list))

                                    
    if __name__ == "__main__":
        train()

**Comment by seiriosPlus** [comment]
把ADAM换成SGD看看速度会不会有提升

**Comment by LihangLiu** [comment]
换乘SGD的话is_sparse=True确实快了很多。
所以是adam跟is_sparse冲突了？

**Comment by seiriosPlus** [comment]
后续会针对ADAM在sparse的情况下的计算再进行优化， 可以尝试其他几个跟ADAM效果接近的optimizer。

**Comment by velconia** [comment]
> 所以是adam跟is_sparse冲突了？

Adam Op 现在会:
1. 先将 Sparse 的 Tensor 转换为 Dense 的 Tensor;
2. 针对Embedding表的所有字段都进行更新动量的计算以达到数学上正确的结果;

后续讨论后会针对是否一定要保证数学上正确这块再做优化;

**Comment by velconia** [comment]
@LihangLiu 请问你的实际参数有多大呢？如果实际参数比较小，设置is_sparse=True是没有太大意义的，反而会变慢

**Comment by LihangLiu** [comment]
实际参数挺多的，有embed_size=200k的特征。

**Comment by velconia** [comment]
@LihangLiu 我们稍后会先针对CPU优化一下算法复杂度，然后开启Op内多线程的能力，用以彻底解决这个问题

**Comment by LihangLiu** [comment]
非常感谢！修改之后还是可以用gpu来训练是吧？还是说只是优化了cpu版本？

**Comment by velconia** [comment]
和 @jacquesqiao  讨论之后, 结论如下:

1. 增加lazy_mode, 只更新有梯度的参数部分，CPU GPU通用
    优点：实现快速；
    缺点：会对效果有影响；
https://github.com/PaddlePaddle/Paddle/pull/14889

2. 实现Op内多线程，提升CPU下的性能；
    优点：对效果无影响，提速明显；
    缺点：改动较大；
https://github.com/PaddlePaddle/Paddle/pull/14890

3. 针对没有梯度的参数和有梯度的参数进行不同方法的更新, 提升CPU下的性能；
    优点：对效果无影响，实现快速；
    缺点：提速不明显；
https://github.com/PaddlePaddle/Paddle/pull/14891

**Comment by velconia** [comment]
鉴于仅有 https://github.com/PaddlePaddle/Paddle/pull/14889 这个PR对GPU部分产生影响，这个issue可以先close了，请问还有别的问题吗？
